rewrite synthesizer, reachability analysis

arifogel opened the pull request:
- Fix #867 where synthesizer caused persistent mutations to cached Configurations
- cut down functionality to known used/working features
- redo reachability AST node design
- extract functionality from AST nodes into visitors
- move default drop fibrow generation to bdp from synthesizer
- move rule generation from synthesizer into new State classes
- extend CommonUtil.computeIpOwners to support new synthesizer configuration options
- fix swallowed NPE when deserializing answer to outliers question
  - surfaced due to ref differences caused by fix for #867
- update refs due to fix for #867

<!-- Reviewable:start -->
---
This change is [<img src="https://reviewable.io/review_button.svg" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/batfish/batfish/902)
<!-- Reviewable:end -->

anothermattbrown commented:
Reviewed 1 of 189 files at r1.
Review status: 1 of 190 files reviewed at latest revision, 1 unresolved discussion.

---

*[projects/batfish-common-protocol/src/main/java/org/batfish/common/util/CommonUtil.java, line 209 at r1](https://reviewable.io:443/reviews/batfish/batfish/902#-L4iQvwDmkS4FPmPKanL:-L4iQvwDmkS4FPmPKanM:b-b0y916) ([raw file](https://github.com/batfish/batfish/blob/573c9fdc7a7e20f0ba2b6443101a8f34dd27bf5b/projects/batfish-common-protocol/src/main/java/org/batfish/common/util/CommonUtil.java#L209)):*
> ```Java
>       Map<String, Configuration> configurations,
>       boolean excludeInactive,
>       Map<String, Set<String>> disabledInterfaces,
> ```

This is further complicating an already quite complex method. Is this hot code? If not, it may be better to optimize for simplicity rather than performance by filtering out disabled owners as a separate postprocessing step.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/batfish/batfish/902)*
<!-- Sent from Reviewable.io -->

progwriter commented:
Making my way thought this in passes, so far only very minor comments and a plea for javadoc/comments.

---

Reviewed 44 of 189 files at r1, 1 of 1 files at r2, 2 of 2 files at r3.
Review status: 47 of 190 files reviewed at latest revision, 4 unresolved discussions.

---

*[projects/batfish/src/main/java/org/batfish/z3/SynthesizerInput.java, line 191 at r3](https://reviewable.io:443/reviews/batfish/batfish/902#-L4lUO3GvUSY6YM7MR9N:-L4lUO3GvUSY6YM7MR9O:bd9xbwj) ([raw file](https://github.com/batfish/batfish/blob/512a12eac694fa398bf21bc6eaeef7dd2c1f6a18/projects/batfish/src/main/java/org/batfish/z3/SynthesizerInput.java#L191)):*
> ```Java
>   }
> 
>   private Map<String, Map<String, Map<Integer, BooleanExpr>>> computeAclConditions() {
> ```

I feel like this method (and similar methods below) need to be Javadoc'ed and/or commented. I stared at this for a while and still don't fully grok what's being returned.

---

*[projects/batfish/src/main/java/org/batfish/z3/expr/BooleanExprVisitor.java, line 1 at r3](https://reviewable.io:443/reviews/batfish/batfish/902#-L4lRHywdSOMctp2riJQ:-L4lRHyxdDqiRGhpuIAK:bxrjzhw) ([raw file](https://github.com/batfish/batfish/blob/512a12eac694fa398bf21bc6eaeef7dd2c1f6a18/projects/batfish/src/main/java/org/batfish/z3/expr/BooleanExprVisitor.java#L1)):*
> ```Java
> package org.batfish.z3.expr;
> ```

Should probably moved to the `visitors` package for consistency

---

*[projects/batfish/src/main/java/org/batfish/z3/expr/visitors/ExprPrinter.java, line 57 at r3](https://reviewable.io:443/reviews/batfish/batfish/902#-L4lPp_SieQ4VDAC33qV:-L4lPp_SieQ4VDAC33qW:bl0csw6) ([raw file](https://github.com/batfish/batfish/blob/512a12eac694fa398bf21bc6eaeef7dd2c1f6a18/projects/batfish/src/main/java/org/batfish/z3/expr/visitors/ExprPrinter.java#L57)):*
> ```Java
>   }
> 
>   private boolean isComplexExpr(Expr expr) {
> ```

Just a nitpick, but this can probably be `static`? 

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/batfish/batfish/902#-:-L4lTyS_3Js_u-YD1dZC:bafzkr4)*
<!-- Sent from Reviewable.io -->

anothermattbrown commented:
Reviewed 160 of 189 files at r1.
Review status: all files reviewed at latest revision, 14 unresolved discussions.

---

*[projects/batfish/src/main/java/org/batfish/z3/expr/RangeMatchExpr.java, line 10 at r1](https://reviewable.io:443/reviews/batfish/batfish/902#-L4lp_Ye-ErLKIEF1Ntx:-L4lp_Ye-ErLKIEF1Nty:b6avoja) ([raw file](https://github.com/batfish/batfish/blob/573c9fdc7a7e20f0ba2b6443101a8f34dd27bf5b/projects/batfish/src/main/java/org/batfish/z3/expr/RangeMatchExpr.java#L10)):*
> ```Java
> public class RangeMatchExpr extends BooleanExpr {
> 
>   public static BooleanExpr greaterThanOrEqualTo(HeaderField bv, long lb, int numBits) {
> ```

Z3's built-in is_bv_uge is probably more efficient

---

*[projects/batfish/src/main/java/org/batfish/z3/expr/RangeMatchExpr.java, line 50 at r1](https://reviewable.io:443/reviews/batfish/batfish/902#-L4lpvx2LYfEc32ZfoH3:-L4lpvx2LYfEc32ZfoH4:bkiok4z) ([raw file](https://github.com/batfish/batfish/blob/573c9fdc7a7e20f0ba2b6443101a8f34dd27bf5b/projects/batfish/src/main/java/org/batfish/z3/expr/RangeMatchExpr.java#L50)):*
> ```Java
>   }
> 
>   public static BooleanExpr lessThanOrEqualTo(HeaderField bv, long lb, int numBits) {
> ```

similar comment to above

---

*[projects/batfish/src/main/java/org/batfish/z3/state/AclDeny.java, line 30 at r1](https://reviewable.io:443/reviews/batfish/batfish/902#-L4iaTOcNyyBfOevjT4d:-L4iaTOcNyyBfOevjT4e:b7j27c9) ([raw file](https://github.com/batfish/batfish/blob/573c9fdc7a7e20f0ba2b6443101a8f34dd27bf5b/projects/batfish/src/main/java/org/batfish/z3/state/AclDeny.java#L30)):*
> ```Java
>           .stream()
>           .flatMap(
>               e -> {
> ```

Maybe pull this out and give it a descriptive name (either as a local variable or a method)?

---

*[projects/batfish/src/main/java/org/batfish/z3/state/AclDeny.java, line 69 at r1](https://reviewable.io:443/reviews/batfish/batfish/902#-L4iahP-W2n-nwgpw6a3:-L4iahP-W2n-nwgpw6a4:benpsw9) ([raw file](https://github.com/batfish/batfish/blob/573c9fdc7a7e20f0ba2b6443101a8f34dd27bf5b/projects/batfish/src/main/java/org/batfish/z3/state/AclDeny.java#L69)):*
> ```Java
>           .stream()
>           .flatMap(
>               e -> {
> ```

Same comment as above.

---

*[projects/batfish/src/main/java/org/batfish/z3/state/AclLineMatch.java, line 31 at r1](https://reviewable.io:443/reviews/batfish/batfish/902#-L4iatEBirIQga08VrFn:-L4iatEBirIQga08VrFo:b-7mc15d) ([raw file](https://github.com/batfish/batfish/blob/573c9fdc7a7e20f0ba2b6443101a8f34dd27bf5b/projects/batfish/src/main/java/org/batfish/z3/state/AclLineMatch.java#L31)):*
> ```Java
>           .stream()
>           .flatMap(
>               e -> {
> ```

name?

---

*[projects/batfish/src/main/java/org/batfish/z3/state/AclLineMatch.java, line 37 at r1](https://reviewable.io:443/reviews/batfish/batfish/902#-L4iauCdseT9XBwGIpHq:-L4iauCdseT9XBwGIpHr:b-7mc15d) ([raw file](https://github.com/batfish/batfish/blob/573c9fdc7a7e20f0ba2b6443101a8f34dd27bf5b/projects/batfish/src/main/java/org/batfish/z3/state/AclLineMatch.java#L37)):*
> ```Java
>                     .stream()
>                     .flatMap(
>                         e2 -> {
> ```

name?

---

*[projects/batfish/src/main/java/org/batfish/z3/state/AclLineMatch.java, line 43 at r1](https://reviewable.io:443/reviews/batfish/batfish/902#-L4iavE0-jNm9cKAbidS:-L4iavE0-jNm9cKAbidT:b-7mc15d) ([raw file](https://github.com/batfish/batfish/blob/573c9fdc7a7e20f0ba2b6443101a8f34dd27bf5b/projects/batfish/src/main/java/org/batfish/z3/state/AclLineMatch.java#L43)):*
> ```Java
>                               .stream()
>                               .map(
>                                   e3 -> {
> ```

name?

---

*[projects/batfish/src/main/java/org/batfish/z3/state/State.java, line 21 at r1](https://reviewable.io:443/reviews/batfish/batfish/902#-L4ih9H3B4a-G30O-UeV:-L4ih9H3B4a-G30O-UeW:bgvdg12) ([raw file](https://github.com/batfish/batfish/blob/573c9fdc7a7e20f0ba2b6443101a8f34dd27bf5b/projects/batfish/src/main/java/org/batfish/z3/state/State.java#L21)):*
> ```Java
>       extends BooleanExpr {
> 
>     public static final List<Expr> VARIABLES =
> ```

Hard-coding this here makes this unnecessarily rigid, and will cause problems for e.g. solving for a variable number of destinations.

Why should states have to know what the variables are at all?

---

*[projects/batfish/src/main/java/org/batfish/z3/state/StateParameterization.java, line 3 at r1](https://reviewable.io:443/reviews/batfish/batfish/902#-L4iedQ0RwueLuut179-:-L4iedQ0RwueLuut1790:b8d2pgt) ([raw file](https://github.com/batfish/batfish/blob/573c9fdc7a7e20f0ba2b6443101a8f34dd27bf5b/projects/batfish/src/main/java/org/batfish/z3/state/StateParameterization.java#L3)):*
> ```Java
> package org.batfish.z3.state;
> 
> public interface StateParameterization<T> {
> ```

What is the purpose of this? It seems to hold the fields of each state type. Why not have the state hold their own fields?

---

*[projects/batfish/src/main/java/org/batfish/z3/state/Transition.java, line 8 at r1](https://reviewable.io:443/reviews/batfish/batfish/902#-L4liKA7j_e99yCbL21d:-L4liKA7j_e99yCbL21e:b2xwxrd) ([raw file](https://github.com/batfish/batfish/blob/573c9fdc7a7e20f0ba2b6443101a8f34dd27bf5b/projects/batfish/src/main/java/org/batfish/z3/state/Transition.java#L8)):*
> ```Java
> 
> public interface Transition<T> {
>   List<RuleExpr> generate(SynthesizerInput input);
> ```

I'd prefer to make this a visitor pattern accept method, and move generate into a visitor. This will put all the generation logic into one class (making it easier to understand and maintain), and will make it easier to add other operations.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/batfish/batfish/902)*
<!-- Sent from Reviewable.io -->

anothermattbrown commented:
Review status: all files reviewed at latest revision, 14 unresolved discussions.

---

*[projects/batfish/src/main/java/org/batfish/z3/expr/visitors/ExprPrinter.java, line 57 at r3](https://reviewable.io:443/reviews/batfish/batfish/902#-L4lPp_SieQ4VDAC33qV:-L4ltSYn5JYdYg70J0uh:b7ztt0a) ([raw file](https://github.com/batfish/batfish/blob/512a12eac694fa398bf21bc6eaeef7dd2c1f6a18/projects/batfish/src/main/java/org/batfish/z3/expr/visitors/ExprPrinter.java#L57)):*
<details><summary><i>Previously, progwriter (Victor Heorhiadi) wrote…</i></summary><blockquote>

Just a nitpick, but this can probably be `static`? 
</blockquote></details>

Or make a this method a visitor. That way, if we add a new Expr the typechecker will force us to add a case for isComplexExpr. This will silently default it to NOT be complex, which is error-prone.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/batfish/batfish/902)*
<!-- Sent from Reviewable.io -->

anothermattbrown commented:
Review status: all files reviewed at latest revision, 14 unresolved discussions.

---

*[projects/batfish/src/main/java/org/batfish/z3/SynthesizerInput.java, line 191 at r3](https://reviewable.io:443/reviews/batfish/batfish/902#-L4lUO3GvUSY6YM7MR9N:-L4ltswNzGRCvAxmX0DO:bb6mt9b) ([raw file](https://github.com/batfish/batfish/blob/512a12eac694fa398bf21bc6eaeef7dd2c1f6a18/projects/batfish/src/main/java/org/batfish/z3/SynthesizerInput.java#L191)):*
<details><summary><i>Previously, progwriter (Victor Heorhiadi) wrote…</i></summary><blockquote>

I feel like this method (and similar methods below) need to be Javadoc'ed and/or commented. I stared at this for a while and still don't fully grok what's being returned.
</blockquote></details>

:+1:

---

*[projects/batfish/src/main/java/org/batfish/z3/expr/BooleanExprVisitor.java, line 1 at r3](https://reviewable.io:443/reviews/batfish/batfish/902#-L4lRHywdSOMctp2riJQ:-L4ltzDxZ818I1u8GWJE:bb6mt9b) ([raw file](https://github.com/batfish/batfish/blob/512a12eac694fa398bf21bc6eaeef7dd2c1f6a18/projects/batfish/src/main/java/org/batfish/z3/expr/BooleanExprVisitor.java#L1)):*
<details><summary><i>Previously, progwriter (Victor Heorhiadi) wrote…</i></summary><blockquote>

Should probably moved to the `visitors` package for consistency
</blockquote></details>

:+1:

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/batfish/batfish/902)*
<!-- Sent from Reviewable.io -->

arifogel commented:
Review status: 129 of 191 files reviewed at latest revision, 14 unresolved discussions.

---

*[projects/batfish-common-protocol/src/main/java/org/batfish/common/util/CommonUtil.java, line 209 at r1](https://reviewable.io:443/reviews/batfish/batfish/902#-L4iQvwDmkS4FPmPKanL:-L4ldMpAPxoGHKmIyNyz:b-33f439) ([raw file](https://github.com/batfish/batfish/blob/573c9fdc7a7e20f0ba2b6443101a8f34dd27bf5b/projects/batfish-common-protocol/src/main/java/org/batfish/common/util/CommonUtil.java#L209)):*
<details><summary><i>Previously, anothermattbrown (Matt Brown) wrote…</i></summary><blockquote>

This is further complicating an already quite complex method. Is this hot code? If not, it may be better to optimize for simplicity rather than performance by filtering out disabled owners as a separate postprocessing step.
</blockquote></details>

I moved all filtering to `SynthesizerInput`constructor. This function and all transition generators now just iterate over the `enabledConfigurations`, `enabledInterfaces`, etc. without needing to add filter steps everywhere.

---

*[projects/batfish/src/main/java/org/batfish/z3/SynthesizerInput.java, line 191 at r3](https://reviewable.io:443/reviews/batfish/batfish/902#-L4lUO3GvUSY6YM7MR9N:-L4lgsGn0bz88ikSGGwS:b600wk3) ([raw file](https://github.com/batfish/batfish/blob/512a12eac694fa398bf21bc6eaeef7dd2c1f6a18/projects/batfish/src/main/java/org/batfish/z3/SynthesizerInput.java#L191)):*
<details><summary><i>Previously, progwriter (Victor Heorhiadi) wrote…</i></summary><blockquote>

I feel like this method (and similar methods below) need to be Javadoc'ed and/or commented. I stared at this for a while and still don't fully grok what's being returned.
</blockquote></details>

I added javadocs to the getters.

---

*[projects/batfish/src/main/java/org/batfish/z3/expr/visitors/ExprPrinter.java, line 57 at r3](https://reviewable.io:443/reviews/batfish/batfish/902#-L4lPp_SieQ4VDAC33qV:-L4lfGHdTtYCBHx8DHNF:bb6un1j) ([raw file](https://github.com/batfish/batfish/blob/512a12eac694fa398bf21bc6eaeef7dd2c1f6a18/projects/batfish/src/main/java/org/batfish/z3/expr/visitors/ExprPrinter.java#L57)):*
<details><summary><i>Previously, progwriter (Victor Heorhiadi) wrote…</i></summary><blockquote>

Just a nitpick, but this can probably be `static`? 
</blockquote></details>

Done

---

*[projects/batfish/src/main/java/org/batfish/z3/expr/BooleanExprVisitor.java, line 1 at r3](https://reviewable.io:443/reviews/batfish/batfish/902#-L4lRHywdSOMctp2riJQ:-L4lfATc02DlJFOVHOma:b-ql8z78) ([raw file](https://github.com/batfish/batfish/blob/512a12eac694fa398bf21bc6eaeef7dd2c1f6a18/projects/batfish/src/main/java/org/batfish/z3/expr/BooleanExprVisitor.java#L1)):*
<details><summary><i>Previously, progwriter (Victor Heorhiadi) wrote…</i></summary><blockquote>

Should probably moved to the `visitors` package for consistency
</blockquote></details>

Done for the 3 visitors I created.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/batfish/batfish/902)*
<!-- Sent from Reviewable.io -->

anothermattbrown commented:
Review status: 129 of 191 files reviewed at latest revision, 14 unresolved discussions.

---

*[projects/batfish-common-protocol/src/main/java/org/batfish/common/util/CommonUtil.java, line 209 at r1](https://reviewable.io:443/reviews/batfish/batfish/902#-L4iQvwDmkS4FPmPKanL:-L4lvQ9SbmmE1_3-cVYb:bt3e1we) ([raw file](https://github.com/batfish/batfish/blob/573c9fdc7a7e20f0ba2b6443101a8f34dd27bf5b/projects/batfish-common-protocol/src/main/java/org/batfish/common/util/CommonUtil.java#L209)):*
<details><summary><i>Previously, arifogel (Ari Fogel) wrote…</i></summary><blockquote>

I moved all filtering to `SynthesizerInput`constructor. This function and all transition generators now just iterate over the `enabledConfigurations`, `enabledInterfaces`, etc. without needing to add filter steps everywhere.
</blockquote></details>

Does that mean the `excludeInactive` and `disabledInterfaces` parameters are going away?

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/batfish/batfish/902)*
<!-- Sent from Reviewable.io -->

anothermattbrown commented:
Reviewed 2 of 192 files at r1, 24 of 65 files at r4.
Review status: 151 of 191 files reviewed at latest revision, 14 unresolved discussions, some commit checks failed.

---

*[projects/batfish-common-protocol/src/main/java/org/batfish/common/util/CommonUtil.java, line 209 at r1](https://reviewable.io:443/reviews/batfish/batfish/902#-L4iQvwDmkS4FPmPKanL:-L4lx3rJdc7sQ4-Vxipg:b9d6lzu) ([raw file](https://github.com/batfish/batfish/blob/573c9fdc7a7e20f0ba2b6443101a8f34dd27bf5b/projects/batfish-common-protocol/src/main/java/org/batfish/common/util/CommonUtil.java#L209)):*
<details><summary><i>Previously, anothermattbrown (Matt Brown) wrote…</i></summary><blockquote>

Does that mean the `excludeInactive` and `disabledInterfaces` parameters are going away?
</blockquote></details>

Ahh, sorry reviewable was showing me the wrong version for some reason.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/batfish/batfish/902)*
<!-- Sent from Reviewable.io -->

arifogel commented:
Review status: 151 of 191 files reviewed at latest revision, 14 unresolved discussions, some commit checks failed.

---

*[projects/batfish/src/main/java/org/batfish/z3/expr/visitors/ExprPrinter.java, line 57 at r3](https://reviewable.io:443/reviews/batfish/batfish/902#-L4lPp_SieQ4VDAC33qV:-L4lyjojU4qsSuX3mq9t:b-9afss2) ([raw file](https://github.com/batfish/batfish/blob/512a12eac694fa398bf21bc6eaeef7dd2c1f6a18/projects/batfish/src/main/java/org/batfish/z3/expr/visitors/ExprPrinter.java#L57)):*
<details><summary><i>Previously, arifogel (Ari Fogel) wrote…</i></summary><blockquote>

Done
</blockquote></details>

Made it a visitor.

---

*[projects/batfish/src/main/java/org/batfish/z3/state/StateParameterization.java, line 3 at r1](https://reviewable.io:443/reviews/batfish/batfish/902#-L4iedQ0RwueLuut179-:-L4lz2R6iuWLEPfysxFs:bpr61mf) ([raw file](https://github.com/batfish/batfish/blob/573c9fdc7a7e20f0ba2b6443101a8f34dd27bf5b/projects/batfish/src/main/java/org/batfish/z3/state/StateParameterization.java#L3)):*
<details><summary><i>Previously, anothermattbrown (Matt Brown) wrote…</i></summary><blockquote>

What is the purpose of this? It seems to hold the fields of each state type. Why not have the state hold their own fields?
</blockquote></details>

The plan was to be able to be able to turn on and off bitvector encoding (rather than encoding into the name) of  individual `StateParameter` variables by `StateParameter.Type` without having to know about each `State` that uses it.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/batfish/batfish/902)*
<!-- Sent from Reviewable.io -->

arifogel commented:
Review status: 149 of 191 files reviewed at latest revision, 14 unresolved discussions.

---

*[projects/batfish/src/main/java/org/batfish/z3/state/AclDeny.java, line 30 at r1](https://reviewable.io:443/reviews/batfish/batfish/902#-L4iaTOcNyyBfOevjT4d:-L4m-z3-W-BFlZuIJ9Rp:b-896fix) ([raw file](https://github.com/batfish/batfish/blob/573c9fdc7a7e20f0ba2b6443101a8f34dd27bf5b/projects/batfish/src/main/java/org/batfish/z3/state/AclDeny.java#L30)):*
<details><summary><i>Previously, anothermattbrown (Matt Brown) wrote…</i></summary><blockquote>

Maybe pull this out and give it a descriptive name (either as a local variable or a method)?
</blockquote></details>

Done.

---

*[projects/batfish/src/main/java/org/batfish/z3/state/AclDeny.java, line 69 at r1](https://reviewable.io:443/reviews/batfish/batfish/902#-L4iahP-W2n-nwgpw6a3:-L4m-zX3BOK9mZoitoM0:b-896fix) ([raw file](https://github.com/batfish/batfish/blob/573c9fdc7a7e20f0ba2b6443101a8f34dd27bf5b/projects/batfish/src/main/java/org/batfish/z3/state/AclDeny.java#L69)):*
<details><summary><i>Previously, anothermattbrown (Matt Brown) wrote…</i></summary><blockquote>

Same comment as above.
</blockquote></details>

Done.

---

*[projects/batfish/src/main/java/org/batfish/z3/state/AclLineMatch.java, line 31 at r1](https://reviewable.io:443/reviews/batfish/batfish/902#-L4iatEBirIQga08VrFn:-L4m0CuP6rAhLU9W54ST:b-896fix) ([raw file](https://github.com/batfish/batfish/blob/573c9fdc7a7e20f0ba2b6443101a8f34dd27bf5b/projects/batfish/src/main/java/org/batfish/z3/state/AclLineMatch.java#L31)):*
<details><summary><i>Previously, anothermattbrown (Matt Brown) wrote…</i></summary><blockquote>

name?
</blockquote></details>

Done.

---

*[projects/batfish/src/main/java/org/batfish/z3/state/AclLineMatch.java, line 37 at r1](https://reviewable.io:443/reviews/batfish/batfish/902#-L4iauCdseT9XBwGIpHq:-L4m0DHQ8NYBj1PJcMLw:b-896fix) ([raw file](https://github.com/batfish/batfish/blob/573c9fdc7a7e20f0ba2b6443101a8f34dd27bf5b/projects/batfish/src/main/java/org/batfish/z3/state/AclLineMatch.java#L37)):*
<details><summary><i>Previously, anothermattbrown (Matt Brown) wrote…</i></summary><blockquote>

name?
</blockquote></details>

Done.

---

*[projects/batfish/src/main/java/org/batfish/z3/state/AclLineMatch.java, line 43 at r1](https://reviewable.io:443/reviews/batfish/batfish/902#-L4iavE0-jNm9cKAbidS:-L4m0DdwK88-x9zzRz-Q:b-896fix) ([raw file](https://github.com/batfish/batfish/blob/573c9fdc7a7e20f0ba2b6443101a8f34dd27bf5b/projects/batfish/src/main/java/org/batfish/z3/state/AclLineMatch.java#L43)):*
<details><summary><i>Previously, anothermattbrown (Matt Brown) wrote…</i></summary><blockquote>

name?
</blockquote></details>

Done.

---

*[projects/batfish/src/main/java/org/batfish/z3/state/State.java, line 21 at r1](https://reviewable.io:443/reviews/batfish/batfish/902#-L4ih9H3B4a-G30O-UeV:-L4m0c9LqTX55WP3Ng3d:bwinl4r) ([raw file](https://github.com/batfish/batfish/blob/573c9fdc7a7e20f0ba2b6443101a8f34dd27bf5b/projects/batfish/src/main/java/org/batfish/z3/state/State.java#L21)):*
<details><summary><i>Previously, anothermattbrown (Matt Brown) wrote…</i></summary><blockquote>

Hard-coding this here makes this unnecessarily rigid, and will cause problems for e.g. solving for a variable number of destinations.

Why should states have to know what the variables are at all?
</blockquote></details>

I'd like to defer this until I implement the thing I was talking about in my reply about reason for `Parameterization`

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/batfish/batfish/902)*
<!-- Sent from Reviewable.io -->

arifogel commented:
Review status: 135 of 196 files reviewed at latest revision, 9 unresolved discussions.

---

*[projects/batfish/src/main/java/org/batfish/z3/expr/visitors/ExprPrinter.java, line 36 at r7](https://reviewable.io:443/reviews/batfish/batfish/902#-L4qr-zXN4c_oioAs2vc:-L4sYPrzDYu4q4O9-L42:b9mij66) ([raw file](https://github.com/batfish/batfish/blob/1cf11611ec0dff979a8acb9946f9ed5dce23d1a0/projects/batfish/src/main/java/org/batfish/z3/expr/visitors/ExprPrinter.java#L36)):*
<details><summary><i>Previously, anothermattbrown (Matt Brown) wrote…</i></summary><blockquote>

Self-contained, so may as well factor it out.
</blockquote></details>

Elaborate, please.

---

*[projects/batfish/src/main/java/org/batfish/z3/expr/visitors/ExprPrinter.java, line 88 at r7](https://reviewable.io:443/reviews/batfish/batfish/902#-L4qqg6PZkRjfQmAxLTg:-L4sZ_UBXK2JcIIomWik:b-9koke5) ([raw file](https://github.com/batfish/batfish/blob/1cf11611ec0dff979a8acb9946f9ed5dce23d1a0/projects/batfish/src/main/java/org/batfish/z3/expr/visitors/ExprPrinter.java#L88)):*
<details><summary><i>Previously, anothermattbrown (Matt Brown) wrote…</i></summary><blockquote>

Presumably we got here through a call to expr.accept(this), so this would probably cause an infinite loop. Might be better to throw an exception.
</blockquote></details>

Got rid of the method entirely, as it never should have existed.

---

*[projects/batfish/src/main/java/org/batfish/z3/state/Transition.java, line 8 at r1](https://reviewable.io:443/reviews/batfish/batfish/902#-L4liKA7j_e99yCbL21d:-L4sZhiM_8un8ws_HhEF:b-1tswb8) ([raw file](https://github.com/batfish/batfish/blob/573c9fdc7a7e20f0ba2b6443101a8f34dd27bf5b/projects/batfish/src/main/java/org/batfish/z3/state/Transition.java#L8)):*
<details><summary><i>Previously, anothermattbrown (Matt Brown) wrote…</i></summary><blockquote>

I'd prefer to make this a visitor pattern accept method, and move generate into a visitor. This will put all the generation logic into one class (making it easier to understand and maintain), and will make it easier to add other operations.
</blockquote></details>

Gone in refactoring. For now, I removed ability to disable transitions. Leaving for another PR.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/batfish/batfish/902)*
<!-- Sent from Reviewable.io -->

arifogel commented:
Review status: 129 of 196 files reviewed at latest revision, 9 unresolved discussions.

---

*[projects/batfish/src/main/java/org/batfish/z3/expr/RangeMatchExpr.java, line 10 at r1](https://reviewable.io:443/reviews/batfish/batfish/902#-L4lp_Ye-ErLKIEF1Ntx:-L4shGI_nmuITYss6xlC:bo41a3p) ([raw file](https://github.com/batfish/batfish/blob/573c9fdc7a7e20f0ba2b6443101a8f34dd27bf5b/projects/batfish/src/main/java/org/batfish/z3/expr/RangeMatchExpr.java#L10)):*
<details><summary><i>Previously, anothermattbrown (Matt Brown) wrote…</i></summary><blockquote>

Z3's built-in is_bv_uge is probably more efficient
</blockquote></details>

Probably, but I'd like to defer to another PR

---

*[projects/batfish/src/main/java/org/batfish/z3/expr/RangeMatchExpr.java, line 50 at r1](https://reviewable.io:443/reviews/batfish/batfish/902#-L4lpvx2LYfEc32ZfoH3:-L4shJrnB-h4JBBe8tKB:bo41a3p) ([raw file](https://github.com/batfish/batfish/blob/573c9fdc7a7e20f0ba2b6443101a8f34dd27bf5b/projects/batfish/src/main/java/org/batfish/z3/expr/RangeMatchExpr.java#L50)):*
<details><summary><i>Previously, anothermattbrown (Matt Brown) wrote…</i></summary><blockquote>

similar comment to above
</blockquote></details>

Probably, but I'd like to defer to another PR

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/batfish/batfish/902)*
<!-- Sent from Reviewable.io -->

arifogel commented:
Review status: 129 of 196 files reviewed at latest revision, 9 unresolved discussions.

---

*[projects/batfish/src/main/java/org/batfish/z3/state/AclDeny.java, line 30 at r1](https://reviewable.io:443/reviews/batfish/batfish/902#-L4iaTOcNyyBfOevjT4d:-L4shXkRx3RTfpbQPsTK:b-7wisfe) ([raw file](https://github.com/batfish/batfish/blob/573c9fdc7a7e20f0ba2b6443101a8f34dd27bf5b/projects/batfish/src/main/java/org/batfish/z3/state/AclDeny.java#L30)):*
<details><summary><i>Previously, anothermattbrown (Matt Brown) wrote…</i></summary><blockquote>

Sorry, I meant the entire lambda expression. How about a comment at each lambda abstraction?
</blockquote></details>

Everything has been moved, and should look simpler. If you still feel this way, please make similar comments in `DefaultTransitionGenerator`

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/batfish/batfish/902)*
<!-- Sent from Reviewable.io -->

anothermattbrown commented:
Reviewed 3 of 192 files at r1, 2 of 61 files at r9.
Review status: 129 of 196 files reviewed at latest revision, 4 unresolved discussions.

---

*[projects/batfish/src/main/java/org/batfish/z3/expr/visitors/ExprPrinter.java, line 36 at r7](https://reviewable.io:443/reviews/batfish/batfish/902#-L4qr-zXN4c_oioAs2vc:-L4vuB3ckEFy7zi_BAUT:bxcpbn8) ([raw file](https://github.com/batfish/batfish/blob/1cf11611ec0dff979a8acb9946f9ed5dce23d1a0/projects/batfish/src/main/java/org/batfish/z3/expr/visitors/ExprPrinter.java#L36)):*
<details><summary><i>Previously, arifogel (Ari Fogel) wrote…</i></summary><blockquote>

Elaborate, please.
</blockquote></details>

Doesn't seem to be a reason to have this as an inner class. Thought it might be cleaner to have it in its own file. Doesn't really matter though.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/batfish/batfish/902)*
<!-- Sent from Reviewable.io -->

arifogel commented:
Review status: 129 of 196 files reviewed at latest revision, 1 unresolved discussion.

---

*[projects/batfish/src/main/java/org/batfish/z3/expr/visitors/ExprPrinter.java, line 36 at r7](https://reviewable.io:443/reviews/batfish/batfish/902#-L4qr-zXN4c_oioAs2vc:-L4wHA6XaH7oav7ses9C:b-bj9cb0) ([raw file](https://github.com/batfish/batfish/blob/1cf11611ec0dff979a8acb9946f9ed5dce23d1a0/projects/batfish/src/main/java/org/batfish/z3/expr/visitors/ExprPrinter.java#L36)):*
<details><summary><i>Previously, anothermattbrown (Matt Brown) wrote…</i></summary><blockquote>

Doesn't seem to be a reason to have this as an inner class. Thought it might be cleaner to have it in its own file. Doesn't really matter though.
</blockquote></details>

I moved it

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/batfish/batfish/902)*
<!-- Sent from Reviewable.io -->

progwriter commented:
OK to me. Deferring to @anothermattbrown for actual approval :) 

---

Reviewed 6 of 65 files at r4, 2 of 2 files at r7, 55 of 61 files at r9, 12 of 12 files at r10, 1 of 2 files at r11.
Review status: 196 of 197 files reviewed at latest revision, 1 unresolved discussion.

---



*Comments from [Reviewable](https://reviewable.io:443/reviews/batfish/batfish/902#-:-L4wHVwE0suO8fIbwz0e:benh8lz)*
<!-- Sent from Reviewable.io -->

anothermattbrown commented:
<img class="emoji" title=":lgtm:" alt=":lgtm:" align="absmiddle" src="https://reviewable.io/lgtm.png" height="20" width="61"/>

---

Reviewed 1 of 2 files at r11.
Review status: all files reviewed at latest revision, all discussions resolved.

---



*Comments from [Reviewable](https://reviewable.io:443/reviews/batfish/batfish/902#-:-L5A49d-lVF5q-jcp9qj:bnfp4nl)*
<!-- Sent from Reviewable.io -->

dhalperi commented:
Review status: all files reviewed at latest revision, 1 unresolved discussion, all commit checks successful.

---

*[a discussion](https://reviewable.io:443/reviews/batfish/batfish/902#-L5A7LW-nd70TGaPFKKz:-L5A7LW-nd70TGaPFKL-:b-bdj0hl) (no related file):*
You mentioned that this rewrite caught a lot of bugs, yet I see zero new tests. Please add tests.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/batfish/batfish/902)*
<!-- Sent from Reviewable.io -->

arifogel commented:
Review status: 171 of 204 files reviewed at latest revision, 1 unresolved discussion.

---

*[a discussion](https://reviewable.io:443/reviews/batfish/batfish/902#-L5A7LW-nd70TGaPFKKz:-L5C7q9zfMAs4fxeRxVY:bkfo0zp) (no related file):*
<details><summary><i>Previously, dhalperi (Dan Halperin) wrote…</i></summary><blockquote>

You mentioned that this rewrite caught a lot of bugs, yet I see zero new tests. Please add tests.
</blockquote></details>

FYI I added some initial tests for `SynthesizerInput`. I'm not done with that class, but feel free to review what I have so far.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/batfish/batfish/902)*
<!-- Sent from Reviewable.io -->

dhalperi commented:
Reviewed 103 of 195 files at r1, 1 of 1 files at r2, 2 of 2 files at r3, 8 of 68 files at r4, 43 of 62 files at r9, 1 of 12 files at r10, 21 of 38 files at r12, 11 of 36 files at r14, 17 of 22 files at r15, 7 of 8 files at r16, 3 of 3 files at r17.
Review status: all files reviewed at latest revision, 3 unresolved discussions.

---

*[projects/batfish/src/main/java/org/batfish/z3/expr/DeclareVarStatement.java, line 6 at r17](https://reviewable.io:443/reviews/batfish/batfish/902#-L5M-Y1Z647BVpG3ZS0E:-L5M-Y1Z647BVpG3ZS0F:b-5oqygl) ([raw file](https://github.com/batfish/batfish/blob/63cc569ab400f68600adf63a76ec051b81b88953/projects/batfish/src/main/java/org/batfish/z3/expr/DeclareVarStatement.java#L6)):*
> ```Java
> import org.batfish.z3.HeaderField;
> 
> public class DeclareVarStatement extends Statement {
> ```

OK

High-level comment: some contextual Javadoc on every public class would be useful.

---

*[projects/batfish/src/main/java/org/batfish/z3/expr/ExtractExpr.java, line 8 at r17](https://reviewable.io:443/reviews/batfish/batfish/902#-L5M-lVVOi6nbEIdEIHT:-L5M-lVVOi6nbEIdEIHU:bnmcuoi) ([raw file](https://github.com/batfish/batfish/blob/63cc569ab400f68600adf63a76ec051b81b88953/projects/batfish/src/main/java/org/batfish/z3/expr/ExtractExpr.java#L8)):*
> ```Java
> import org.batfish.z3.expr.visitors.IntExprVisitor;
> 
> public class ExtractExpr extends IntExpr {
> ```

OK

(The javadoc may be more useful for non-trivial things -- presumably this corresponds to extracting some bits from an IP address?)

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/batfish/batfish/902)*
<!-- Sent from Reviewable.io -->

dhalperi commented:
Update on coverage, if anyone is interested.

```
org.batfish.z3	10% (3/28)	42% (72/171)	30% (349/1142)
org.batfish.z3.expr	76% (23/30)	42% (78/185)	49% (327/662)
org.batfish.z3.expr.visitors	50% (3/6)	18% (24/128)	29% (134/450)
org.batfish.z3.state	90% (60/66)	89% (191/213)	90% (320/354)
org.batfish.z3.state.visitors	100% (2/2)	91% (111/121)	97% (583/599)
```

Generated using 
![Screen Shot 2018-02-15 at 1.55.57 PM.png](https://files.reviewableusercontent.io/images/reviewable/526415/CRlXVvYxA8RxDU4rK_9xbtLw/Screen%20Shot%202018-02-15%20at%201.55.57%20PM.png)

---







*Comments from [Reviewable](https://reviewable.io:443/reviews/batfish/batfish/902#-:-L5QSjUpqWRzspE0aTmU:b-g95imf)*
<!-- Sent from Reviewable.io -->

anothermattbrown commented:
Very cool. Is there a way to get automated coverage reports, e.g. from travis?

---

Reviewed 1 of 68 files at r4, 5 of 39 files at r12, 2 of 37 files at r14, 1 of 23 files at r15, 12 of 17 files at r18, 9 of 9 files at r19.
Review status: all files reviewed at latest revision, 4 unresolved discussions.

---

*[projects/batfish/src/main/java/org/batfish/z3/SynthesizerInput.java, line 21 at r19](https://reviewable.io:443/reviews/batfish/batfish/902#-L5QTsNk6YbsZR_Kdr0w:-L5QTsNk6YbsZR_Kdr0x:b6tm45r) ([raw file](https://github.com/batfish/batfish/blob/9eee7a5bb9afd3a675399d66072a125c4d063732/projects/batfish/src/main/java/org/batfish/z3/SynthesizerInput.java#L21)):*
> ```Java
>    * reasonably encounter (e.g. ACLs assigned to interfaces).
>    */
>   Map<String, Map<String, Map<Integer, LineAction>>> getAclActions();
> ```

This type isn't very informative. Would a single map with keys like the following work?
class AclActionsKey {
  String hostname;
  String aclName;
  Integer lineNumber;
}

Or is there a specific need for the hierarchy?

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/batfish/batfish/902#-:-L5QT4EqG4T2YnYj4v0Y:b-71vuf0)*
<!-- Sent from Reviewable.io -->

arifogel commented:
Review status: all files reviewed at latest revision, 3 unresolved discussions.

---

*[projects/batfish/src/main/java/org/batfish/z3/SynthesizerInput.java, line 21 at r19](https://reviewable.io:443/reviews/batfish/batfish/902#-L5QTsNk6YbsZR_Kdr0w:-L5ROOZtPCqISNs_j2CP:b32e5v) ([raw file](https://github.com/batfish/batfish/blob/9eee7a5bb9afd3a675399d66072a125c4d063732/projects/batfish/src/main/java/org/batfish/z3/SynthesizerInput.java#L21)):*
<details><summary><i>Previously, anothermattbrown (Matt Brown) wrote…</i></summary><blockquote>

This type isn't very informative. Would a single map with keys like the following work?
class AclActionsKey {
  String hostname;
  String aclName;
  Integer lineNumber;
}

Or is there a specific need for the hierarchy?
</blockquote></details>

The hierarchy is for performance. Any changes are out of scope for this PR.

---

*[projects/batfish/src/main/java/org/batfish/z3/expr/DeclareVarStatement.java, line 6 at r17](https://reviewable.io:443/reviews/batfish/batfish/902#-L5M-Y1Z647BVpG3ZS0E:-L5ROVwqROECfDZqVGdn:b-nwx54s) ([raw file](https://github.com/batfish/batfish/blob/63cc569ab400f68600adf63a76ec051b81b88953/projects/batfish/src/main/java/org/batfish/z3/expr/DeclareVarStatement.java#L6)):*
<details><summary><i>Previously, dhalperi (Dan Halperin) wrote…</i></summary><blockquote>

OK

High-level comment: some contextual Javadoc on every public class would be useful.
</blockquote></details>

I agree, but please let's defer to another PR.

---

*[projects/batfish/src/main/java/org/batfish/z3/expr/ExtractExpr.java, line 8 at r17](https://reviewable.io:443/reviews/batfish/batfish/902#-L5M-lVVOi6nbEIdEIHT:-L5ROoFQc3aB_1C3KROZ:bzgtt8c) ([raw file](https://github.com/batfish/batfish/blob/63cc569ab400f68600adf63a76ec051b81b88953/projects/batfish/src/main/java/org/batfish/z3/expr/ExtractExpr.java#L8)):*
<details><summary><i>Previously, dhalperi (Dan Halperin) wrote…</i></summary><blockquote>

OK

(The javadoc may be more useful for non-trivial things -- presumably this corresponds to extracting some bits from an IP address?)
</blockquote></details>

Added doc to that effect.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/batfish/batfish/902)*
<!-- Sent from Reviewable.io -->

dhalperi commented:
Reviewed 8 of 17 files at r18.
Review status: all files reviewed at latest revision, 2 unresolved discussions.

---

*[projects/batfish/src/main/java/org/batfish/z3/expr/DeclareVarStatement.java, line 6 at r17](https://reviewable.io:443/reviews/batfish/batfish/902#-L5M-Y1Z647BVpG3ZS0E:-L5SDzusygSozizzpOno:b-xzrqeh) ([raw file](https://github.com/batfish/batfish/blob/63cc569ab400f68600adf63a76ec051b81b88953/projects/batfish/src/main/java/org/batfish/z3/expr/DeclareVarStatement.java#L6)):*
<details><summary><i>Previously, arifogel (Ari Fogel) wrote…</i></summary><blockquote>

I agree, but please let's defer to another PR.
</blockquote></details>

Ok, but let's set it as a principle going forward: new classes get documentation on public APIs.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/batfish/batfish/902)*
<!-- Sent from Reviewable.io -->
